ql/src/java/org/apache/hadoop/hive/ql/parse/TezCompiler.java

```diff
@@ -1483,17 +1489,64 @@ private void removeSemijoinsParallelToMapJoin(OptimizeTezProcContext procCtx)
       }
     }
 
-    if (semijoins.size() > 0) {
-      for (ReduceSinkOperator rs : semijoins.keySet()) {
-        if (LOG.isDebugEnabled()) {
-          LOG.debug("Semijoin optimization with parallel edge to map join. Removing semijoin "
-              + OperatorUtils.getOpNamePretty(rs) + " - " + OperatorUtils.getOpNamePretty(semijoins.get(rs)));
+    if (!procCtx.conf.getBoolVar(ConfVars.TEZ_DYNAMIC_SEMIJOIN_REDUCTION_FOR_MAPJOIN)) {
+      if (semijoins.size() > 0) {
```

Comment:
You may still need to get rid of the SJ branches for the map join operators here if `HIVE_MAPJOIN_PROBEDECODE_ENABLED`? Or do you need to keep them?

Reply:
Hey @jcamachor  -- in this case I just followed the logic [above](https://github.com/apache/hive/pull/952/files#diff-228178321e4c1df7f045fe181de2fffaL1450) where we seems to bail-out when 
TEZ_DYNAMIC_SEMIJOIN_REDUCTION_FOR_MAPJOIN is enabled.

Reply:
Probably I am missing part of the mechanism to enable your optimization. It seems you are skipping the removal of SJ, so you are adding the context AND keeping those branches in the plan. Is the intention to keep both optimizations in these cases? Or is it because there is some dependency between them? I thought for MJ, the intention was to keep only your new optimization, that's maybe where my misunderstanding is coming from.

Reply:
Hey Jesus, SJ removal and MJ probedecode are independent from my point of view, and thats what I was trying to achieve here -- when TEZ_DYNAMIC_SEMIJOIN_REDUCTION_FOR_MAPJOIN is not enabled SJ removal is triggered (similar to existing code) and when HIVE_OPTIMIZE_SCAN_PROBEDECODE is enabled we do the extra the MJ to TS selection.
I am not sure there is a case where these two optimizations could affect each other (?) but currently they are completely independent.